pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go

```diff
@@ -149,42 +147,64 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	if err != nil {
 		applog.Error(err, "[Handle GenerateApplicationConfiguration]")
 		app.Status.SetConditions(errorCondition("Built", err))
-		r.Recorder.Event(app, corev1.EventTypeWarning, velatypes.ReasonFailedRender, fmt.Sprintf(velatypes.MessageFailedRender, err))
+		r.Recorder.Event(app, event.Warning(velatypes.ReasonFailedRender, err))
 		return handler.handleErr(err)
 	}
 
 	err = handler.handleResourceTracker(ctx, comps, ac)
 	if err != nil {
 		applog.Error(err, "[Handle resourceTracker]")
 		app.Status.SetConditions(errorCondition("Handle resourceTracker", err))
-		r.Recorder.Event(app, corev1.EventTypeWarning, velatypes.ReasonFailedRender, fmt.Sprintf(velatypes.MessageFailedRender, err))
+		r.Recorder.Event(app, event.Warning(velatypes.ReasonFailedRender, err))
 		return handler.handleErr(err)
 	}
 
 	// pass the App label and annotation to ac except some app specific ones
 	oamutil.PassLabelAndAnnotation(app, ac)
 
 	app.Status.SetConditions(readyCondition("Built"))
-	r.Recorder.Event(app, corev1.EventTypeNormal, velatypes.ReasonRendered, velatypes.MessageRendered)
+	r.Recorder.Event(app, event.Normal(velatypes.ReasonRendered, velatypes.MessageRendered))
 	applog.Info("apply application revision & component to the cluster")
 	// apply application revision & component to the cluster
 	if err := handler.apply(ctx, appRev, ac, comps); err != nil {
 		applog.Error(err, "[Handle apply]")
 		app.Status.SetConditions(errorCondition("Applied", err))
-		r.Recorder.Event(app, corev1.EventTypeWarning, velatypes.ReasonFailedApply, fmt.Sprintf(velatypes.MessageFailedApply, err))
+		r.Recorder.Event(app, event.Warning(velatypes.ReasonFailedApply, err))
 		return handler.handleErr(err)
 	}
 
+	// if rolloutPlan is nil, user should create a AppRollout object to rollout the application
+	if !handler.inplace && handler.app.Spec.RolloutPlan != nil {
+		res, err := handler.handleRollout(ctx)
+		if err != nil {
+			applog.Error(err, "[handle rollout]")
+			app.Status.SetConditions(errorCondition("Rollout", err))
+			r.Recorder.Event(app, event.Warning(velatypes.ReasonFailedRollout, err))
+			return handler.handleErr(err)
+		}
+		// skip health check and garbage collection if rollout have not finished
+		// start next reconcile immediately
+		if res.Requeue || res.RequeueAfter > 0 {
+			app.Status.Phase = common.ApplicationRollingOut
+			return res, r.UpdateStatus(ctx, app)
+		}
+
+		// there is no need reconcile immediately, that means the rollout operation have finished
+		r.Recorder.Event(app, event.Normal(velatypes.ReasonRollout, velatypes.MessageRollout))
+		app.Status.SetConditions(readyCondition("Rollout"))
+		applog.Info("rollout finished")
+	}
+
```

Comment:
```suggestion
// The following logic will be skipped if rollout have not finished
```

We should also add annotation here.